File: ui/tui/usage_fetcher.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

const (
	usageEndpoint = "https://chatgpt.com/backend-api/wham/usage"
	usageCacheTTL = 60 * time.Second
)

type usageWindow struct {
	Type      string  `json:"type"`
	Percent   float64 `json:"percentage"`
	Remaining float64 `json:"remaining"`
	ResetAtMs int64   `json:"resetAtMs"`
}

type usageData struct {
	FetchedAt    int64        `json:"fetchedAt"`
	PlanType     string       `json:"planType"`
	Windows      []usageWindow `json:"windows"`
	Allowed      bool         `json:"allowed"`
	LimitReached bool         `json:"limitReached"`
}

func loadCachedUsage(stateDir string, profile string, now time.Time) (*usageData, bool) {
	p := filepath.Join(stateDir, "cache", "usage", profile+".json")
	raw, err := os.ReadFile(p)
	if err != nil {
		return nil, false
	}
	var u usageData
	if err := json.Unmarshal(raw, &u); err != nil {
		return nil, false
	}
	if u.FetchedAt <= 0 {
		return nil, false
	}
	fetched := time.UnixMilli(u.FetchedAt)
	if now.Sub(fetched) > usageCacheTTL {
		return nil, false
	}
	return &u, true
}

func saveCachedUsage(stateDir string, profile string, u *usageData) {
	if u == nil {
		return
	}
	dir := filepath.Join(stateDir, "cache", "usage")
	_ = os.MkdirAll(dir, 0o755)
	p := filepath.Join(dir, profile+".json")
	b, err := json.MarshalIndent(u, "", "  ")
	if err != nil {
		return
	}
	_ = os.WriteFile(p, append(b, '\n'), 0o644)
}

func fetchUsage(accessToken string, accountID string) (*usageData, error) {
	if accessToken == "" {
		return nil, errors.New("missing access token")
	}

	client := &http.Client{
		Timeout: 10 * time.Second,
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{MinVersion: tls.VersionTLS12},
		},
	}

	req, err := http.NewRequest("GET", usageEndpoint, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "MyLLMWorkbench/1.0")
	if accountID != "" {
		req.Header.Set("ChatGPT-Account-Id", accountID)
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		_, _ = io.Copy(io.Discard, resp.Body)
		return nil, errors.New("non-200 from usage endpoint")
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var parsed map[string]any
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, err
	}

	u := parseUsageResponse(parsed)
	return &u, nil
}

func parseUsageResponse(data map[string]any) usageData {
	u := usageData{
		FetchedAt:    time.Now().UnixMilli(),
		PlanType:     asString(data["plan_type"]),
		Windows:      []usageWindow{},
		Allowed:      true,
		LimitReached: false,
	}

	rl, ok := data["rate_limit"].(map[string]any)
	if ok {
		if allowed, ok := rl["allowed"].(bool); ok {
			u.Allowed = allowed
		}
		if lr, ok := rl["limit_reached"].(bool); ok {
			u.LimitReached = lr
		}
		if pw, ok := rl["primary_window"].(map[string]any); ok {
			used := asFloat(pw["used_percent"])
			resetAt := asInt64(pw["reset_at"])
			u.Windows = append(u.Windows, usageWindow{
				Type:      "5h",
				Percent:   used,
				Remaining: 100 - used,
				ResetAtMs: resetAt * 1000,
			})
		}
		if sw, ok := rl["secondary_window"].(map[string]any); ok {
			used := asFloat(sw["used_percent"])
			resetAt := asInt64(sw["reset_at"])
			u.Windows = append(u.Windows, usageWindow{
				Type:      "weekly",
				Percent:   used,
				Remaining: 100 - used,
				ResetAtMs: resetAt * 1000,
			})
		}
	}

	return u
}

func asString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func asFloat(v any) float64 {
	switch t := v.(type) {
	case float64:
		return t
	case int:
		return float64(t)
	case int64:
		return float64(t)
	default:
		return 0
	}
}

func asInt64(v any) int64 {
	switch t := v.(type) {
	case float64:
		return int64(t)
	case int:
		return int64(t)
	case int64:
		return t
	default:
		return 0
	}
}

File: ui/tui/usageFetcher.js
/**
 * Usage Fetcher for Codex OAuth accounts
 * Fetches usage data from ChatGPT backend API
 */

import https from 'node:https';
import { join } from 'node:path';
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs';

const USAGE_ENDPOINT = 'https://chatgpt.com/backend-api/wham/usage';
const CACHE_TTL_MS = 60 * 1000; // 1 minute cache

/**
 * Fetch usage data for a profile
 * @param {string} accessToken - OAuth access token
 * @param {string} accountId - ChatGPT account ID (optional)
 * @returns {Promise<Object|null>} Usage data or null on error
 */
export async function fetchUsage(accessToken, accountId = null) {
  return new Promise((resolve) => {
    const url = new URL(USAGE_ENDPOINT);

    const options = {
      hostname: url.hostname,
      path: url.pathname,
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Accept': 'application/json',
        'User-Agent': 'MyLLMWorkbench/1.0',
      },
      timeout: 10000,
    };

    if (accountId) {
      options.headers['ChatGPT-Account-Id'] = accountId;
    }

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        if (res.statusCode !== 200) {
          resolve(null);
          return;
        }
        try {
          const json = JSON.parse(data);
          resolve(parseUsageResponse(json));
        } catch {
          resolve(null);
        }
      });
    });

    req.on('error', () => resolve(null));
    req.on('timeout', () => {
      req.destroy();
      resolve(null);
    });

    req.end();
  });
}

/**
 * Parse usage response into a standardized format
 * @param {Object} data - Raw API response
 * @returns {Object} Parsed usage data
 */
function parseUsageResponse(data) {
  const result = {
    fetchedAt: Date.now(),
    planType: data.plan_type || null,
    windows: [],
    credits: null,
    allowed: true,
    limitReached: false,
  };

  // Parse rate_limit structure (official ChatGPT API format)
  if (data.rate_limit) {
    result.allowed = data.rate_limit.allowed !== false;
    result.limitReached = data.rate_limit.limit_reached === true;

    // Primary window (5h)
    if (data.rate_limit.primary_window) {
      const pw = data.rate_limit.primary_window;
      result.windows.push({
        type: '5h',
        percentage: pw.used_percent ?? 0,
        remaining: 100 - (pw.used_percent ?? 0),
        windowSeconds: pw.limit_window_seconds || 18000,
        resetAfterSeconds: pw.reset_after_seconds,
        resetAtMs: pw.reset_at ? pw.reset_at * 1000 : null, // Convert Unix timestamp to ms
      });
    }

    // Secondary window (weekly)
    if (data.rate_limit.secondary_window) {
      const sw = data.rate_limit.secondary_window;
      result.windows.push({
        type: 'weekly',
        percentage: sw.used_percent ?? 0,
        remaining: 100 - (sw.used_percent ?? 0),
        windowSeconds: sw.limit_window_seconds || 604800,
        resetAfterSeconds: sw.reset_after_seconds,
        resetAtMs: sw.reset_at ? sw.reset_at * 1000 : null, // Convert Unix timestamp to ms
      });
    }
  }

  // Parse credits
  if (data.credits) {
    result.credits = {
      balance: parseFloat(data.credits.balance) || 0,
      hasCredits: data.credits.has_credits === true,
      unlimited: data.credits.unlimited === true,
      approxLocalMessages: data.credits.approx_local_messages || [0, 0],
      approxCloudMessages: data.credits.approx_cloud_messages || [0, 0],
    };
  }

  // Fallback: parse older formats
  if (result.windows.length === 0) {
    // Try usage_windows array format
    if (data.usage_windows || data.windows) {
      const windows = data.usage_windows || data.windows || [];
      for (const w of (Array.isArray(windows) ? windows : [])) {
        result.windows.push({
          type: w.type || w.name || 'unknown',
          percentage: w.used_percent ?? w.percentage ?? 0,
          remaining: 100 - (w.used_percent ?? w.percentage ?? 0),
          resetAtMs: w.reset_at ? (w.reset_at > 9999999999 ? w.reset_at : w.reset_at * 1000) : null,
        });
      }
    }

    // Try direct percentage fields
    if (data.rate_limit_5h !== undefined) {
      result.windows.push({
        type: '5h',
        percentage: Math.round(data.rate_limit_5h * 100),
        remaining: Math.round((1 - data.rate_limit_5h) * 100),
        resetAtMs: data.reset_5h ? new Date(data.reset_5h).getTime() : null,
      });
    }
    if (data.rate_limit_weekly !== undefined) {
      result.windows.push({
        type: 'weekly',
        percentage: Math.round(data.rate_limit_weekly * 100),
        remaining: Math.round((1 - data.rate_limit_weekly) * 100),
        resetAtMs: data.reset_weekly ? new Date(data.reset_weekly).getTime() : null,
      });
    }
  }

  return result;
}

/**
 * Load cached usage data
 * @param {string} stateDir - State directory path
 * @param {string} profileName - Profile name
 * @returns {Object|null} Cached data or null
 */
export function loadCachedUsage(stateDir, profileName) {
  const cachePath = join(stateDir, 'cache', 'usage', `${profileName}.json`);
  try {
    if (!existsSync(cachePath)) return null;
    const data = JSON.parse(readFileSync(cachePath, 'utf8'));
    // Check if cache is still valid
    if (data.fetchedAt && (Date.now() - data.fetchedAt) < CACHE_TTL_MS) {
      return data;
    }
    return null;
  } catch {
    return null;
  }
}

/**
 * Save usage data to cache
 * @param {string} stateDir - State directory path
 * @param {string} profileName - Profile name
 * @param {Object} data - Usage data to cache
 */
export function saveCachedUsage(stateDir, profileName, data) {
  const cacheDir = join(stateDir, 'cache', 'usage');
  const cachePath = join(cacheDir, `${profileName}.json`);
  try {
    mkdirSync(cacheDir, { recursive: true });
    writeFileSync(cachePath, JSON.stringify(data, null, 2) + '\n', 'utf8');
  } catch {}
}

/**
 * Fetch usage for all profiles in the pool
 * @param {Object} oauthPool - OAuth pool object
 * @param {string} stateDir - State directory path
 * @returns {Promise<Map<string, Object>>} Map of profile name to usage data
 */
export async function fetchAllUsage(oauthPool, stateDir) {
  const results = new Map();

  if (!oauthPool || !oauthPool.profiles) {
    return results;
  }

  const profiles = Object.entries(oauthPool.profiles);

  // Fetch in parallel with rate limiting (max 2 concurrent)
  const batchSize = 2;
  for (let i = 0; i < profiles.length; i += batchSize) {
    const batch = profiles.slice(i, i + batchSize);
    const promises = batch.map(async ([name, profile]) => {
      // Check cache first
      const cached = loadCachedUsage(stateDir, name);
      if (cached) {
        return [name, cached];
      }

      // Fetch fresh data
      const accessToken = profile.accessToken || profile.access_token;
      const accountId = profile.accountId || profile.account_id;

      if (!accessToken) {
        return [name, null];
      }

      const usage = await fetchUsage(accessToken, accountId);
      if (usage) {
        saveCachedUsage(stateDir, name, usage);
      }
      return [name, usage];
    });

    const batchResults = await Promise.all(promises);
    for (const [name, usage] of batchResults) {
      if (usage) {
        results.set(name, usage);
      }
    }
  }

  return results;
}

/**
 * Format reset time for display
 * @param {number|string} resetAt - Reset timestamp or ISO string
 * @returns {string} Formatted string like "2h 30m" or "Jan 25"
 */
export function formatResetTime(resetAt) {
  if (!resetAt) return '?';

  const resetMs = typeof resetAt === 'number' ? resetAt : new Date(resetAt).getTime();
  const now = Date.now();
  const diff = resetMs - now;

  if (diff <= 0) return 'now';

  const seconds = Math.floor(diff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);

  if (days > 0) {
    return `${days}d${hours % 24}h`;
  }

  if (hours > 0) {
    return `${hours}h${minutes % 60}m${seconds % 60}s`;
  }

  if (minutes > 0) {
    return `${minutes}m${seconds % 60}s`;
  }

  return `${seconds}s`;
}
